src/PostgREST/App.hs

```diff
@@ -109,96 +106,104 @@ app dbStructure conf apiRequest =
-                    ("Content-Location",
-                      "/" <> toS (qiName qi) <>
-                        if BS.null canonical then "" else "?" <> toS canonical
-                    )
-                  ] (toS body)
-
-        (ActionCreate, TargetIdent qi@(QualifiedIdentifier _ table), Just payload@(PayloadJSON rows)) ->
+                  (status, contentRange) = rangeHeader queryTotal tableTotal
+                  canonical = iCanonicalQS apiRequest
+              return $
+                if contentType == CTSingularJSON && queryTotal /= 1
+                  then singularityError (toInteger queryTotal)
+                  else responseLBS status
+                    [toHeader contentType, contentRange,
+                      ("Content-Location",
+                        "/" <> toS (qiName qi) <>
+                          if BS.null canonical then "" else "?" <> toS canonical
+                      )
+                    ] (toS body)
+
+        (ActionCreate, TargetIdent (QualifiedIdentifier _ table), Just payload@(PayloadJSON rows)) ->
           case mutateSqlParts of
             Left errorResponse -> return errorResponse
             Right (sq, mq) -> do
               let isSingle = (==1) $ V.length rows
-              when (not isSingle && iPreferSingular apiRequest) $
-                HT.sql [P6.q| DO $$
-                          BEGIN RAISE EXCEPTION cardinality_violation
-                          USING MESSAGE =
-                            'plurality=singular specified, but more than one object would be inserted';
-                          END $$;
-                        |]
-              let pKeys = map pkName $ filter (filterPk schema table) allPrKeys -- would it be ok to move primary key detection in the query itself?
-              let stm = createWriteStatement qi sq mq isSingle (iPreferRepresentation apiRequest) pKeys (contentType == CTTextCSV) payload
-              row <- H.query payload stm
-              let (_, _, fs, body) = extractQueryResult row
-                  headers = catMaybes [
-                      if null fs
-                        then Nothing
-                        else Just (hLocation, "/" <> toS table <> renderLocationFields fs)
-                    , if iPreferRepresentation apiRequest == Full
-                        then Just $ toHeader contentType
-                        else Nothing
-                    , Just . contentRangeH 1 0 $
-                        toInteger <$> if shouldCount then Just (V.length rows) else Nothing
-                    ]
-
-              return . responseLBS status201 headers $
-                if iPreferRepresentation apiRequest == Full
-                  then toS body else ""
-
-        (ActionUpdate, TargetIdent qi, Just payload) ->
+              if contentType == CTSingularJSON
+                 && not isSingle
+                 && iPreferRepresentation apiRequest == Full
+               then do
+                  HT.sql "ROLLBACK;"
```

Comment:
Here we are in the context of a hasql transaction abstraction, if you do rollback here, won't the abstraction try to do a `commit` after this thing and maybe "poison" the connection?
Is this this thing tested?
I mean, the rollback might succeed, but will the following request/transaction succeed?

to reiterate, this is what i think will happen on the postgresql side, will it work without issues?

```sql
-- first request
begin; --hasql transaction
main query;
rollback; -- the line above
commit; --hasql transaction

-- second request
begin; -- is the connection in a broken state here because of the rollback followed by a commit?
...
```